taichi/codegen/llvm/codegen_llvm.cpp

```diff
@@ -2823,21 +2823,29 @@ void TaskCodeGenLLVM::visit(FuncCallStmt *stmt) {
          llvm::ConstantInt::get(*llvm_context, llvm::APInt(32, i, true)), val);
   }
   llvm::Value *result_buffer = nullptr;
-  if (stmt->ret_type->is<PrimitiveType>() &&
-      !stmt->ret_type->is_primitive(PrimitiveTypeID::unknown)) {
-    result_buffer = builder->CreateAlloca(tlctx->get_data_type<uint64>());
-    call("RuntimeContext_set_result_buffer", new_ctx, result_buffer);
-    call(llvm_func, new_ctx);
-    auto *ret_val_u64 = builder->CreateLoad(
+  result_buffer = builder->CreateAlloca(llvm::ArrayType::get(
+      tlctx->get_data_type<uint64>(), stmt->func->rets.size()));
```

Comment:
Wouldn't the process of making a struct type be this:

```c++
std::vector<llvm::Type *> member_types(stmt->func->rets.size());
for (int i = 0; i < stmt->func->rets.size(); i++) {
  member_types[i] = tlctx->get_data_type(stmt->func->rets[i].dt);
}
llvm::StructType *ret_struct_type = llvm::StructType::get(tlctx->get_this_thread_context(), llvm::ArrayRef(member_types));
result_buffer = builder->CreateAlloca(ret_struct_type);
```

And I think the rest of the code should just work after removing all the bitcasts. GEP works the same way in structs iirc

Reply:
But it will become a bit complicated when there are nested structs (if we do not flatten the structs). I would like to add them properly in another PR.